src/cpp/server/DBActiveSessionStorage.cpp

```diff
@@ -0,0 +1,404 @@
+            .withInput(sessionId_);
+         
+         error = connection->execute(updateQuery);
+         if(error)
+         {
+            return Error{"DatabaseException", errc::DBError, "Error while updating properties [ session:"+sessionId_+" properties:"+getKeyString(properties)+" ]", error, ERROR_LOCATION};
+         }
+      }
+      else
+      {
+         database::Query insertQuery = connection->query("INSERT INTO "+kTableName+" ("+kSessionIdColumnName+", "+getKeyString(properties)+") VALUES (:id, "+getValueString(properties)+")")
+            .withInput(sessionId_);
+
+         error = connection->execute(insertQuery);
+         if(error)
+         {
+            return Error{"DatabaseException", errc::DBError, "Error while updating properties [ session:"+sessionId_+" properties:"+getKeyString(properties)+" ]", error, ERROR_LOCATION};
+         }
+      }
+   }
+   return error;
+}
+
+Error DBActiveSessionStorage::destroy()
+{
+   boost::shared_ptr<database::IConnection> connection;
+   Error error = getConnectionOrOverride(&connection);
+
+   if(error)
+   {
+      return error;
+   }
+
+   database::Query query = connection->query("DELETE FROM "+kTableName+" WHERE "+kSessionIdColumnName+" = :id")
+      .withInput(sessionId_);
+   database::Rowset rowset{};
+
+   error = connection->execute(query, rowset);
+
+   if(!error)
+   {
+      return Error{"DatabaseException", errc::DBError, "Error while deleting session metadata [ session:"+sessionId_+" ]", error, ERROR_LOCATION};
+   }
+   else
+   {
+      return error;
+   }
+}
+
+
+Error DBActiveSessionStorage::isValid(bool* pValue)
+{
+   boost::shared_ptr<database::IConnection> connection;
+   Error error = getConnectionOrOverride(&connection);
+
+   if(error)
+   {
+      return error;
+   }
+
+   database::Query query = connection->query("SELECT COUNT(*) FROM "+kTableName+" WHERE "+kSessionIdColumnName+" = :id")
+      .withInput(sessionId_);
+   database::Rowset rowset{};
+
+   *pValue = false;
+
+   error = connection->execute(query, rowset);
+
+   if(error)
+   {
+      return Error{"DatabaseException", errc::DBError, "Error while deleting session metadata [ session:"+sessionId_+" ]", error, ERROR_LOCATION};
+   }
+   database::RowsetIterator iter = rowset.begin();
+
+   // Sanity checking, but should always have 1 row containing the count of rows
+   if(iter != rowset.end())
+   {
+      int count = iter->get<int>("COUNT(*)");
+
+      // ensure one and only one
+      if(count > 1)
+      {
+         return Error{"Too Many Sessions Returned", errc::TooManySessionsReturned, ERROR_LOCATION};
+      }
+      else
+      {
+         if(count == 1)
+         {
+            *pValue = true;
+         }
+      }
+   }
+   return Success();
+}
+
+Error DBActiveSessionStorage::isEmpty(bool* pValue)
+{
+   bool val;
+   Error error = isValid(&val);
```

Comment:
What if the session isn't valid because there are already two entries for the same ID? I think we need a separate check here. Maybe a shared implementation that gets the count and different validators for the count in each method?

Reply:
How about if we add a unique constraint on the active_session_metadata table for the session id?  It's best to avoid this situation and catch those errors on the insert.

Reply:
Since I've already got duplicate row guards in place, Why don't we make swapping that responsiblity a new issue if we want to tackle it. I really want to put a bow on this and spin off any lower priority changes into new issues if we can.